spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+                        # and the message arriving to notify about it
+                        pass
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "pending"
+                ):
+                    other_cmd_info.change_state("cancelled")
+            if mess["type"] != "keepalive" and DISPLAY_STATUS_TABLE:
+                if first:
+                    first = False
+                else:
+                    # overwrite the existing status table
+                    print("\033[2K\033[F" * (4 + len(cmd_infos)))
+                data = [[c.cmd_name, c.state_repr] for c in cmd_infos]
+                header = ["Command", "Status"]
+                msg.table(data, header=header)
+
+        for cmd_info in (c for c in cmd_infos if c.state.name != "cancelled"):
+            msg.divider(cmd_info.cmd_name)
+            if cmd_info.state.name == "not rerun":
+                msg.info(f"Skipping '{cmd_info.cmd_name}': nothing changed")
+            elif cmd_info.console_output is not None:
+                print(cmd_info.console_output)
+        # Occasionally when something has hung there may still be worker processes to tidy up
+        for hung_worker_process in (wp for wp in worker_processes if wp.is_alive()):
+            hung_worker_process.terminate()
+        process_rcs = [c.rc for c in cmd_infos if c.rc is not None]
+        if len(process_rcs) > 0:
+            group_rc = max(process_rcs)
+            if group_rc > 0:
+                msg.fail("A command in the parallel group failed.")
+                sys.exit(group_rc)
+        if any(c for c in cmd_infos if c.state.name in ("hung", "terminated")):
+            msg.fail("Command(s) in the parallel group hung or were terminated.")
+            sys.exit(-1)
+
+
+def _project_run_parallel_cmd(
+    cmd_info: _ParallelCommandInfo,
+    *,
+    dry: bool,
+    current_dir: str,
+    parallel_group_status_queue: Queue,
+) -> None:
+    """Run a single spaCy projects command as a worker process.
+
+    Communicates with the main process via queue messages that consist of
+    dictionaries whose type is determined by the entry 'type'. Possible
+    values of 'type' are 'started', 'completed' and 'keepalive'. Each dictionary
+    type contains different additional fields."""
+    # we can use the command name as a unique log filename because a parallel
+    # group is not allowed to contain the same command more than once
+    log_file_name = os.sep.join(
+        (current_dir, PARALLEL_LOGGING_DIR_NAME, cmd_info.cmd_name + ".log")
+    )
+    file_not_found = False
+    # buffering=0: make sure output is not lost if a subprocess is terminated
+    with open(log_file_name, "wb", buffering=0) as logfile:
+        for os_cmd_index, os_cmd in enumerate(cmd_info.cmd["script"]):
+            command = split_command(os_cmd)
+            if len(command) and command[0] in ("python", "python3"):
```

Comment:
This seems a bit brittle to me (if I understand it correctly). Isn't it possible that the interpreter could also be python3.x?

Reply:
Oh yes I see — I had copied this from `run.py` (ll. 220 onwards) but the original motivation for the feature there was a bit different. It's safe to assume that if someone specifies e.g. `python3.9` they know that's exactly what they want, but in the parallel situation we should still add `-u`.

Do you have any ideas about how we could handle the alias situation?

Reply:
Ah, I see. Since this is the same behavior as in the existing logic I'd ignore it anyway.